src/CitMovie.Data/Repositories/UserScoreRepository.cs

```diff
@@ -0,0 +1,60 @@
+using Microsoft.EntityFrameworkCore;
+
+namespace CitMovie.Data;
+public class UserScoreRepository : IUserScoreRepository
+{
+    private readonly FrameworkContext _frameworkContext;
+    private readonly DataContext _dataContext;
+
+    public UserScoreRepository(FrameworkContext frameworkContext, DataContext dataContext)
+    {
+        _frameworkContext = frameworkContext;
+        _dataContext = dataContext;
+    }
+
+
+    public async Task<IEnumerable<UserScore>> GetUserScoresAsync(int userId, int page, int pageSize, string? mediaType, int? mediaId, string? mediaName)
+    {
+        var userScoresQuery = _frameworkContext.UserScores
+            .Where(us => us.UserId == userId);
+
+        var mediaQuery = _dataContext.Media.AsQueryable();
+        if (!string.IsNullOrEmpty(mediaType))
+        {
+            mediaQuery = mediaQuery.Where(m => m.Type == mediaType);
+        }
+        if (mediaId.HasValue && mediaId > 0)
+        {
+            mediaQuery = mediaQuery.Where(m => m.Id == mediaId);
+        }
+
+        var titlesQuery = _dataContext.Titles.AsQueryable();
+        if (!string.IsNullOrEmpty(mediaName))
+        {
+            titlesQuery = titlesQuery.Where(t => t.Name.Contains(mediaName));
+        }
+
+        var mediaIds = new HashSet<int>(await mediaQuery.Select(m => m.Id).ToListAsync());
+        var titleMediaIds = new HashSet<int>(await titlesQuery.Select(t => t.MediaId).ToListAsync());
+        
+        var filteredUserScores = await userScoresQuery
+            .Where(us => mediaIds.Contains(us.MediaId) && (string.IsNullOrEmpty(mediaName) || titleMediaIds.Contains(us.MediaId)))
+            .Skip(page * pageSize)
+            .Take(pageSize)
+            .ToListAsync();
+
+        return filteredUserScores;
+    }

```

Comment:
Performance will be worse when you call ToList since it forces the query to load its result into memory. I believe that  if you use an IQueryable inside another it kinda works like a WITH statement in plain SQL.

 Btw, DbSet implements IQueryable so there is no need to call AsQueryable.

Also HashSet is generally not used (unless something has changed) just use IEnumerable or List instead. But as mentioned earlier you shouldn't even call tolist as thin moves the query from the db to in memory in the application which performs worse

Reply:
I have a difficult time getting it better, could we look at this together. 

I get the error: _cannot use multiple context instances within a single query execution. Ensure the query uses a single context instance._ when I try not to implement ToList ect.

Reply:
ahh, that must be because you are using the data and framework context. We can look at it together at some other time, lets just get this merged then 👍